Source/DafnyLanguageServer/Workspace/DiagnosticPublisher.cs

```diff
@@ -1,44 +1,109 @@
         // We leave the responsibility to shift the error locations to the LSP clients.
         // Therefore, we do not republish the errors when the document (re-)load was canceled.
         return;
       }
+
+      PublishVerificationStatus(document);
       PublishDocumentDiagnostics(document);
       PublishGhostDiagnostics(document);
     }
 
+    PublishedVerificationStatus FromImplementationTask(IImplementationTask task) {
+      switch (task.CurrentStatus) {
+        case VerificationStatus.Stale: return PublishedVerificationStatus.Stale;
+        case VerificationStatus.Queued:
+          return PublishedVerificationStatus.Queued;
+        case VerificationStatus.Running:
+          return PublishedVerificationStatus.Running;
+        case VerificationStatus.Completed:
+#pragma warning disable VSTHRD002
+          return task.ActualTask.Result.Outcome == ConditionGeneration.Outcome.Correct
+#pragma warning restore VSTHRD002
+            ? PublishedVerificationStatus.Correct
+            : PublishedVerificationStatus.Error;
+        default:
+          throw new ArgumentOutOfRangeException();
+      }
+    }
+
+    static PublishedVerificationStatus Combine(PublishedVerificationStatus first, PublishedVerificationStatus second) {
+      if (first == PublishedVerificationStatus.Error || second == PublishedVerificationStatus.Error) {
+        return PublishedVerificationStatus.Error;
+      }
+
+      return new[] { first, second }.Min();
+    }
+
+    private void PublishVerificationStatus(DafnyDocument document) {
+      var namedVerifiableGroups = document.VerificationTasks.GroupBy(task => task.Implementation.tok.GetLspRange());
+      var namedVerifiableStatusList = namedVerifiableGroups.Select(taskGroup => {
+        var statuses = taskGroup.Select(FromImplementationTask);
+        var status = statuses.Aggregate(Combine);
+        return new NamedVerifiableStatus(taskGroup.Key, status);
+      }).ToList();
+      var notification = new FileVerificationStatus(document.Uri, document.Version, namedVerifiableStatusList);
+
+      if (previouslyVerificationStatus.TryGetValue(document.Uri, out var previous) && previous.Equals(notification)) {
+        Console.Write("");
+      } else {
+        languageServer.TextDocument.SendNotification(VerificationStatusNotification, notification);
+        previouslyVerificationStatus.AddOrUpdate(document.Uri, _ => notification, (_, _) => notification);
+      }
+    }
+
     private void PublishDocumentDiagnostics(DafnyDocument document) {
-      languageServer.TextDocument.PublishDiagnostics(new PublishDiagnosticsParams {
+      var newParams = new PublishDiagnosticsParams {
         Uri = document.Uri,
         Version = document.Version,
         Diagnostics = document.Diagnostics.ToArray(),
-      });
+      };
+      if (previouslyPublishedDiagnostics.TryGetValue(document.Uri, out var previousParams) && previousParams.Diagnostics.Equals(newParams.Diagnostics)) {
```

Comment:
Did you consider `SequenceEquals` instead of `Equals`? Diagnostics are arrays and array equality is only reference-based